src/solvers/smt2/smt2_parser.cpp

```diff
@@ -28,35 +28,48 @@ std::ostream &operator<<(std::ostream &out, const smt2_parsert::smt2_format &f)
 
 void smt2_parsert::command_sequence()
 {
+  PRECONDITION(parenthesis_level == 0);
+
   exit=false;
 
   while(next_token()==OPEN)
   {
-    if(next_token()!=SYMBOL)
+    DATA_INVARIANT(parenthesis_level == 1, "incorrect parenthesis nesting");
+
+    try
     {
-      error() << "expected symbol as command" << eom;
-      return;
-    }
+      if(next_token() != SYMBOL)
+      {
+        error() << "expected symbol as command" << eom;
+        throw smt2_errort();
+      }
 
-    command(buffer);
+      command(buffer);
 
-    if(exit)
-      return;
+      if(exit)
+        return;
 
-    switch(next_token())
+      switch(next_token())
+      {
+      case END_OF_FILE:
+        error() << "expected closing parenthesis at end of command,"
+                   " but got EOF"
+                << eom;
+        throw smt2_errort();
+
+      case CLOSE:
+        // what we expect
+        break;
+
+      default:
+        error() << "expected end of command" << eom;
+        throw smt2_errort();
+      }
+    }
+    catch(smt2_errort)
     {
-    case END_OF_FILE:
-      error() << "expected closing parenthesis at end of command,"
-                 " but got EOF" << eom;
-      return;
-
-    case CLOSE:
-      // what we expect
-      break;
-
-    default:
-      error() << "expected end of command" << eom;
-      return;
+      // consume any tokens till the end of the command
+      skip_to_end_of_list();
```

Comment:
Why?

Reply:
This is to recover cleanly from parse errors.